### DISCUSSION TITLE
Using Route Transforms Outside of Services Builder

### DISCUSSION THREAD

**Discussion by jongallant** [dotnet/yarp#1939] [Q&A] [answered]
Is there no way to use YARP transforms from the request pipeline itself?  Must all transforms be applied at the service builder phase?

I have a requirement where I must insert a header value in a request middleware.  Must I inject it directly into the HttpContext object in this case?  Or can it be done with YARP transforms?

**Comment by Tratcher** [comment, marked as answer]
Middleware can modify the HttpContext as needed, though we recommend transforms as a cleaner option to avoid issues on retry, etc.

What else is the middleware doing that can't be done in a transform?

**Reply by jongallant** [reply]
For example, calling a webservice to pull information on a specific user.  Once data is retrieved, a header is set.

I know this is not a typical case.

**Reply by Tratcher** [reply]
You can do that within a custom transform.
https://microsoft.github.io/reverse-proxy/articles/transforms.html#extensibility
